array_api_tests/test_special_cases.py

```diff
@@ -0,0 +1,1209 @@
+        cond = math.isfinite
+        expr_template = "isfinite({})"
+        if not not_cond:
+            kwargs = {"allow_nan": False, "allow_infinity": False}
+    elif cond_str in "a positive (i.e., greater than ``0``) finite number":
+        cond = lambda i: math.isfinite(i) and i > 0
+        expr_template = "isfinite({}) and {} > 0"
+        if not not_cond:
+            kwargs = {
+                "allow_nan": False,
+                "allow_infinity": False,
+                "min_value": 0,
+                "exclude_min": True,
+            }
+    elif cond_str == "a negative (i.e., less than ``0``) finite number":
+        cond = lambda i: math.isfinite(i) and i < 0
+        expr_template = "isfinite({}) and {} < 0"
+        if not not_cond:
+            kwargs = {
+                "allow_nan": False,
+                "allow_infinity": False,
+                "max_value": 0,
+                "exclude_max": True,
+            }
+    elif cond_str == "positive":
+        cond = lambda i: math.copysign(1, i) == 1
+        expr_template = "copysign(1, {}) == 1"
+        if not not_cond:
+            # We assume (positive) zero is special cased seperately
+            kwargs = {"min_value": 0, "exclude_min": True}
+    elif cond_str == "negative":
+        cond = lambda i: math.copysign(1, i) == -1
+        expr_template = "copysign(1, {}) == -1"
+        if not not_cond:
+            # We assume (negative) zero is special cased seperately
+            kwargs = {"max_value": 0, "exclude_max": True}
+    elif "nonzero finite" in cond_str:
+        cond = lambda i: math.isfinite(i) and i != 0
+        expr_template = "isfinite({}) and {} != 0"
+        if not not_cond:
+            kwargs = {"allow_nan": False, "allow_infinity": False}
+            filter_ = lambda n: n != 0
+    elif cond_str == "an integer value":
+        cond = lambda i: i.is_integer()
+        expr_template = "{}.is_integer()"
+        if not not_cond:
+            from_dtype = integers_from_dtype  # type: ignore
+    elif cond_str == "an odd integer value":
+        cond = lambda i: i.is_integer() and i % 2 == 1
+        expr_template = "{}.is_integer() and {} % 2 == 1"
+        if not not_cond:
+
+            def from_dtype(dtype: DataType, **kw) -> st.SearchStrategy[float]:
+                return integers_from_dtype(dtype, **kw).filter(lambda n: n % 2 == 1)
+
+    else:
+        raise ParseError(cond_str)
+
+    if strat is not None:
+        if (
+            not_cond
+            or len(kwargs) != 0
+            or filter_ is not None
+            or from_dtype is not None
+        ):
+            raise ParseError(cond_str)
+        return cond, expr_template, wrap_strat_as_from_dtype(strat)
+
+    if not_cond:
+        expr_template = f"not {expr_template}"
+        cond = make_not_cond(cond)
+        kwargs = {}
+        filter_ = cond
+    assert kwargs is not None
+    return cond, expr_template, BoundFromDtype(kwargs, filter_, from_dtype)
```

Comment:
comment whats happening in these "stages". also note the difference between using normal strategies and `BoundFromDtype`, and why they're mixed-and-match